source/common/formatter/substitution_formatter.cc

```diff
@@ -105,8 +107,10 @@ std::string FormatterImpl::format(const Http::RequestHeaderMap& request_headers,
   log_line.reserve(256);
 
   for (const FormatterProviderPtr& provider : providers_) {
-    log_line += provider->format(request_headers, response_headers, response_trailers, stream_info,
-                                 local_reply_body);
+    const auto& bit = provider->format(request_headers, response_headers, response_trailers,
+                                       stream_info, local_reply_body);
+    log_line +=
+        bit ? bit.value() : (omit_empty_values_ ? EMPTY_STRING : DefaultUnspecifiedValueString);
```

Comment:
I think you could put 
```
const std::string& empty_value = omit_empty_values_ ? EMPTY_STRING : DefaultUnspecifiedValueString;
```
outside the loop and then just write
```
log_line += bit.value_or(empty_value);
```

Reply:
Didn't know about `value_or`. Neat! thank you. fixed everywhere.